File: Algorithms/Easy/IntersectionOfTwoLinkedLists.go
package easy

import "github.com/parviz-yu/LeetCode/Algorithms/helpers"

// Time complexity: O(n)
// Space complexity: O(1)
func getIntersectionNode(headA, headB *helpers.ListNode) *helpers.ListNode {
	l1, l2 := headA, headB
	for l1 != l2 {
		if l1 != nil {
			l1 = l1.Next
		} else {
			l1 = headB
		}

		if l2 != nil {
			l2 = l2.Next
		} else {
			l2 = headA
		}
	}

	return l1
}

File: Algorithms/Easy/160.Intersection-Of-Two-Linked-Lists/IntersectionOfTwoLinkedLists.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


# Time complexity: O(m + n)
class Solution:
    # Space complexity: O(m)
    def getIntersectionNode_v1(
        self, headA: ListNode, headB: ListNode
    ) -> Optional[ListNode]:
        seen = set()
        while headA:
            if headA not in seen:
                seen.add(headA)

            headA = headA.next

        while headB:
            if headB in seen:
                return headB

            headB = headB.next

        return None

    # Space complexity: O(1)
    def getIntersectionNode_v2(
        self, headA: ListNode, headB: ListNode
    ) -> Optional[ListNode]:
        lenA = self.length(headA)
        lenB = self.length(headB)

        if lenA > lenB:
            headA = self.skipNodes(headA, lenA - lenB)
        elif lenB > lenA:
            headB = self.skipNodes(headB, lenB - lenA)

        while headA and headB:
            if headA == headB:
                return headA

            headA = headA.next
            headB = headB.next

        return None

    def length(self, head: ListNode) -> int:
        curr = head
        count = 0
        while curr:
            count += 1
            curr = curr.next

        return count

    def skipNodes(self, head: ListNode, n: int) -> None:
        for _ in range(n):
            head = head.next

        return head
